src/Compilers/Core/Portable/DiagnosticAnalyzer/AsyncQueue.cs

```diff
@@ -177,6 +177,21 @@ private bool CompleteCore()
             {
                 if (existingWaiters != null)
                 {
+                    // we have waiters and we have data.
+                    // allow some extra time to drain the queue before cancelling
+                    // but do not wait for too long.
+                    SpinWait.SpinUntil(() =>
+                    {
```

Comment:
It was not apparent that we want this to be a general purpose type. It looks very specific to our usage. If that is a case, then, indeed, the delay or whatever ends up the actual fix for the race should be moved to the caller.